lib/recursive-methods.rb

```diff
@@ -1,49 +1,84 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the number handed to the method.
+# Space complexity: O(n) as well
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0 || n == 1
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2), where n is the length of the string passed
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    return s[-1] + reverse(s[0..-2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2), where n is the length of the string passed .. does the replace impact this as well??
+# Space complexity: O(n^2)
 def reverse_inplace(s)
```

Comment:
⚠️ This is not done in place.  You are instead creating new strings with each recursive call.  

Instead consider this method signature

```suggestion
# Time complexity: O(n), where n is the length of the string passed .. does the replace impact this as well??
# Space complexity: O(n)
def reverse_inplace(s, low = 0, high = s.length - 1)
```